Changed jail creation to always create base jails

thorsteneb opened the pull request:
Closes #52

Ornias1993 commented:
Thank you for your PR.
Required checks  and tests for this before merge:
- Test reinstall with currently installed non-base jail
- Test installing a jail from scratch
- Test update with multiple base-jails
- Check all the jails and make sure they don't use any of the "forbiden" folders for basejails (basically the list of /usr folders that are handled differently with basejail)

If thats all okey, it looks good to me :)

thorsteneb commented:
Successful in most things, with one exception. This may be fairly easy to resolve.

root@freenas[~/jailman]# ./jailman.sh -i jacket radarr sonarr lidarr transmission tautulli organizr kms nextcloud mariadb bitwarden
...
Successfully added mount to radarr's fstab
ln: /usr/bin/mono: Read-only file system
Command: ln -s /usr/local/bin/mono /usr/bin/mono failed!
In radarr install.sh:

iocage exec radarr ln -s /usr/local/bin/mono /usr/bin/mono
Whytho?

I am hoping that ln is not required to actually run radarr. I'd be a little shocked if it was.

Same story for the other jails using mono.

thorsteneb commented:
If there are mono shebangs somewhere, and that's why the symlink, following perl's example would be better than symlinking.

```
The /usr/bin/perl symlink has been removed starting with Perl 5.20.
For shebangs, you should either use:

#!/usr/local/bin/perl

or

#!/usr/bin/env perl
```

thorsteneb commented:
Collecting some info from radarr. Instructions at https://github.com/Radarr/Radarr/wiki/Installation reference /usr/local/bin/mono.

Let's find out why that symlink is there, if need be resolve config so it's not needed, and remove the symlink creation from install.sh.

Ornias1993 commented:
It was based on an example made by someone else.
Try cutting it out, run the test again... If it works it works ;)

Ornias1993 commented:
In theory it should be in /usr/local/* ofc, not directly under /usr/

thorsteneb commented:
Am I missing something about sonarr? I've now tried this without base jail, clone jail as in your dev branch, and sonarr will not start:

```
root@sonarr:~ # service sonarr status
sonarr is not running.
root@sonarr:~ # service sonarr start
Starting sonarr.
root@sonarr:~ # service sonarr status
sonarr is running as pid 63422.
root@sonarr:~ # service sonarr status
sonarr is not running.
```

I need to get sonarr running in a clone jail before I can proclaim victory in a base jail. From the logs:

```
20-4-6 19:11:04.6|Info|Bootstrap|Starting Sonarr - /usr/local/share/NzbDrone/NzbDrone.exe - Version 2.0.0.5344
20-4-6 19:11:05.0|Info|AppFolderInfo|Data directory is being overridden to [/config]
20-4-6 19:11:05.1|Info|Router|Application mode: Interactive
20-4-6 19:11:05.2|Info|MigrationLogger|*** Migrating data source=/config/nzbdrone.db;cache size=-10000;datetimekind=Utc;journal mode=Wal;pooling=True;version=3 ***
20-4-6 19:11:05.2|Fatal|ConsoleApp|EPIC FAIL!

[v2.0.0.5344] System.TypeInitializationException: The type initializer for 'System.Data.SQLite.SQLiteFactory' threw an exception. ---> System.DllNotFoundException: libsqlite3.so
  at (wrapper managed-to-native) System.Data.SQLite.UnsafeNativeMethods.sqlite3_config_none(System.Data.SQLite.SQLiteConfigOpsEnum)
```

Where to from here?

Ornias1993 commented:
Interesting the sonarr pkg list seems to be wrong.
Oddly enough it is working fine on my side though... try this:
mono mediainfo sqlite3

thorsteneb commented:
That fixed sonarr, and I can get to it on port 8989. Lidarr and radarr I cannot get to on 8686 and 7878 respectively. radarr service is running; lidarr is not. I just saw your packages-fix branch, will fix locally and test again. Side note: Jackett is missing an install.sh in dev, or in the version I branched from.

thorsteneb commented:
Looking decent. Test runs successful. Multiple jail updates were not working, but that was already inherent in the code, it never worked. Fixed that while I was at it.

Should be good to go.

Ornias1993 commented on config.yml.example:
```diff
@@ -33,13 +33,13 @@ jackett: jackett
 radarr: radarr
-  pkgs: mono
+  pkgs: mono mediainfo sqlite3 libgdiplus
  
 sonarr: sonarr
-  pkgs: mono
+  pkgs: mono mediainfo sqlite3
  
 lidarr: lidarr
-  pkgs: mono
+  pkgs: mono mediainfo sqlite3
```
This should already be on the dev branch now, merged the PR for it right away.
Rebase should clear this.

Ornias1993 commented on jailman.sh:
```diff
@@ -46,7 +46,7 @@ while getopts ":i:r:u:d:g:h" opt
         u ) updatejails=("$OPTARG")
             until [[ $(eval "echo \${$OPTIND}") =~ ^-.* ]] || [ -z $(eval "echo \${$OPTIND}") ]; do
-                updateljails+=($(eval "echo \${$OPTIND}"))
+                updatejails+=($(eval "echo \${$OPTIND}"))
```
This should already be on the dev branch now, merged the PR for it right away.
Rebase should clear this.

Ornias1993 commented:
> Looking decent. Test runs successful. Multiple jail updates were not working, but that was already inherent in the code, it never worked.

Those "multiple jails" are just radarr and lidarr? Just to be clear...
Also didn't work for you with clone jails?

thorsteneb commented:
It was because of the typo "updateljail" instead of "updatejails", fixed with the destroyjails PR.

Rebased on dev and squashed to a single commit. Should be good to go now!

Ornias1993 commented:
Whoops that was more than 24 hours, sorry! :(
